_posts/2023-07-07-python-udf.md

```diff
@@ -0,0 +1,311 @@
+
+1) **Performance.** The function could be executed using the same execution model (e.g., streaming results, beyond-memory execution) of the DBMS, and without any unnecessary transformations.
+2) **Easy Use.** UDFs can be seamlessly integrated into SQL queries, allowing users to leverage the power of SQL to call the functions. This eliminates the need for passing data through a separate database connector and executing external code. The functions can be utilized in various SQL contexts(e.g., subqueries, join conditions).
+3) **Safety.** The sensitive data never leaves the DBMS process. 
+
+There are two main reasons users often refrain from implementing UDFs. 1) There are security concerns associated with UDFs. Since UDFs are custom code created by users and executed within the DBMS process, there is a potential risk of crashing the server. However, when it comes to DuckDB, an embedded database, this concern is mitigated as each analyst runs their own DuckDB process separately. Therefore, the impact on server stability is not a significant worry. 2) The difficulty of implementation is a common deterrent for users. High-Performance UDFs are typically only supported in low-level languages. UDFs in higher-level languages like Python incur significant performance costs. Consequently many users cannot quickly implement their UDFs without investing a significant amount of time in learning a low-level language and understanding the internal details of the DBMS.
+
+DuckDB followed a similar approach. As a DBMS tailored for analytical tasks, performance is a key consideration, leading to the implementation of its core in C++. Consequently, the initial focus of extensibility efforts [was centered around C++](https://www.youtube.com/watch?v=UKo_LQyLTko&ab_channel=DuckDBLabs). However, this  duck is not limited to just waddling; it can also fly. So we are delighted to announce the [recent addition](https://github.com/duckdb/duckdb/pull/7171) of Scalar Python UDFs to DuckDB.
+
+DuckDB provides support for two distinct types of Python UDFs, differing in the Python object used for communication between [DuckDB's native data types](https://duckdb.org/docs/sql/data_types/overview) and the Python process. These communication layers include support for [Python built-in types](https://duckdb.org/docs/sql/data_types/overview) and [PyArrow Tables](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html).
+
+The two approaches exhibit two key differences:
+
+1) **Zero-Copy.** PyArrow Tables leverage our [zero-copy integration with Arrow](https://duckdb.org/2021/12/03/duck-arrow.html), enabling efficient translation of data types to Python-Land with zero-copy cost.
+2) **Vectorization.** PyArrow Table functions operate on a chunk level, processing chunks of data containing up to 2048 rows. This approach maximizes cache locality and leverages vectorization. On the other hand, the built-in types implementation operates on a per-row basis.
```

Comment:
It's somewhat unclear what `the built-in types implementation` refers to. Is it the build-in UDF implementation?